proposals/0000-adding-support-for-swift.md

```diff
@@ -0,0 +1,111 @@
+- Libraries will need to update their documentation and probably some changes to their imports as well (ideally, the API should not change, unless it is related to the Platform specific side of things and the Swift implementation simplifies the existing API for them to consume) and as [previously established](https://github.com/react-native-community/discussions-and-proposals/issues/671) not all libraries adapt and maintainers may not update
+- Users will need to upgrade their templates and the libraries they use to Switch over
+
+## Alternatives
+
+### Leaving things as is
+The overhead of this option is much lower in the short term, and while it creates a bad user experience for the smaller subsection of the user base that want to be able to do more (and library maintainers who would probably benefit from this new-found access), the cost is undeniably lower. The problem is that eventually, Apple will either push for this change directly, or will make it very hard to continue, so unless we take initiative now, We will be forced to it down the road, and in the meantime React Native will have to resort to more and more custom solutions to deal with any APIs that might become unavailable
+
+## Adoption strategy
+
+The core idea is to still be able to deliver feature and changes that users and library maintainers have been asking for while doing the migration. One strategy that comes to mind for achieving this result is as follows:
+- (If needed) Isolating the OSS bits that are user facing from the internals as much as possible (Like AppDelegate and the tests)
+- (If needed) Updating CocoaPods infra
+- (If needed) Communicating with third party libraries to adjust
+- (If needed) Rolling out a new version
+- Migrating the OSS parts from Obj-C -> Swift
+- Updating CLI to support the new files and structures
+- Updating React Native Helper to support people trying to migrate their templates from Obj-C to Swift
+- Updating the docs
+- Communicating with library maintainers if help is needed updating their documentation to support the setup for both Swift and Obj-C for older users
+- Rolling out a new version
+- Moving as much shared logic as possible to C++ and away from Obj-C, separating them as distinctly as possible while adding new features and capabilities and continuing to roll out new versions (**Users should not be affected, and ideally the Library maintainers should not be affected either**)
+- Beginning the transition from Obj-C to Swift from the tests in the internals (**No effect on users or library maintainers**)
+- Moving the rest of the modules one at a time from Obj-C to Swift (**Library maintainers may need to update, if they wish to take advantage of the newer APIs or if we break change an APIs signature, but ideally it should only be in the direction of making it easier to maintain libraries and better APIs**)
+- Updating CocoaPods infra
+- ? (Please add any more steps that come to mind, at the end or anywhere else in the list)
```

Comment:
There is an important step missing.
The New Architecture main language is C++ and it is all conatined in the `ReactCommon` folder.
Most of the code exposed from `ReactCommon` is not configured as a Clang Module. This would make impossible for that code to be consumed by Swift, as [documented here](https://www.swift.org/documentation/cxx-interop/#creating-a-clang-module)

Reply:
Ah, that explains quite a bit. So would we need to configure them as Clang Modules? Is it even possible or would it have some other consequences on other parts of the code base?

Reply:
It is for sure possible.
It will require to write some modulemap files. If that's all it takes, it would not be breaking.
But my gut feeling tells me that we will have to also move some files around, specifically we should move the public headers to an `include` folder. And that would be breaking, unfortunately.

Reply:
Mentioned both the required action and the concern as steps we would need to take under [228b56d](https://github.com/react-native-community/discussions-and-proposals/pull/747/commits/228b56dbbd63f16a1da25d2122fabd32207c2e91)